cryptographic_estimators/MREstimator/mr_algorithm.py

```diff
@@ -0,0 +1,101 @@
+# ****************************************************************************
+# Copyright 2023 Technology Innovation Institute
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+# ****************************************************************************
+
+
+from ..base_algorithm import BaseAlgorithm
+from .mr_problem import MRProblem
+
+
+class MRAlgorithm(BaseAlgorithm):
+    def __init__(self, problem: MRProblem, **kwargs):
+        """
+        Base class for MR algorithms complexity estimator
+
+        INPUT:
+
+        - ``problem`` -- MRProblem object including all necessary parameters
+
+        """
+        super(MRAlgorithm, self).__init__(problem, **kwargs)
+        w = kwargs.get("w", 2.81)
+        theta = kwargs.get("theta", 2)
+        self._name = "BaseMRAlgorithm"
+
+        if  w < 2 or 3 < w:
+            raise ValueError("w must be in the range 2 <= w <= 3")
+
+        if theta is not None and (theta > 2 or theta < 0):
+                raise ValueError("theta must be None or an integer in the range 0 <= theta <= 2")
+
+        self.problem.theta = theta
```

Comment:
That could be confusing: Theta is a property of the problem object, but you are able to set it when creating an algorithm object. It could give the impression that the theta value between algorithms included in the MREstimator can vary, but in fact changing it for one algorithm, will change it for all, as all are based on the same problem object.

Reply:
Unless you see a great advantage in usability I think I would make this property only accessible via the problem object